<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        
        
        <link rel="shortcut icon" href="../../../img/favicon.ico">
        <title>VCF - My Docs</title>
        <link href="../../../css/bootstrap.min.css" rel="stylesheet">
        <link href="../../../css/font-awesome.min.css" rel="stylesheet">
        <link href="../../../css/base.css" rel="stylesheet">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">

        <script src="../../../js/jquery-1.10.2.min.js" defer></script>
        <script src="../../../js/bootstrap.min.js" defer></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
        <script>hljs.initHighlightingOnLoad();</script> 
    </head>

    <body>
        <div class="navbar fixed-top navbar-expand-lg navbar-dark bg-primary">
            <div class="container">
                <a class="navbar-brand" href="../../..">My Docs</a>
                <!-- Expander button -->
                <button type="button" class="navbar-toggler" data-toggle="collapse" data-target="#navbar-collapse">
                    <span class="navbar-toggler-icon"></span>
                </button>

                <!-- Expanded navigation -->
                <div id="navbar-collapse" class="navbar-collapse collapse">
                        <!-- Main navigation -->
                        <ul class="nav navbar-nav">
                            <li class="navitem">
                                <a href="../../.." class="nav-link">Welcome to MkDocs</a>
                            </li>
                            <li class="navitem">
                                <a href="../../../000_todolist/" class="nav-link">000 todolist</a>
                            </li>
                            <li class="navitem">
                                <a href="../../../bookmark/" class="nav-link">Bookmark</a>
                            </li>
                            <li class="navitem">
                                <a href="../../../zzzzz/" class="nav-link">Zzzzz</a>
                            </li>
                            <li class="navitem">
                                <a href="../../../%E5%8D%95%E8%AF%8D/" class="nav-link">单词</a>
                            </li>
                            <li class="navitem">
                                <a href="../../../%E5%B7%A5%E4%BD%9C%E8%AE%B0%E5%BD%95/" class="nav-link">工作记录</a>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">00  <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../00_/000_%E5%9F%BA%E5%9B%A0%E7%BB%93%E6%9E%84/" class="dropdown-item">000 基因结构</a>
</li>
                                    
<li>
    <a href="../../../00_/001_%20CombineVariants/" class="dropdown-item">001  CombineVariants</a>
</li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">01 database <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../01_database/00/" class="dropdown-item">00</a>
</li>
                                    
<li>
    <a href="../../../01_database/01_annovar/" class="dropdown-item">01 annovar</a>
</li>
                                    
<li>
    <a href="../../../01_database/01_ensemble/" class="dropdown-item">01 ensemble</a>
</li>
                                    
<li>
    <a href="../../../01_database/01_vep/" class="dropdown-item">01 vep</a>
</li>
                                    
<li>
    <a href="../../../01_database/02_datasets/" class="dropdown-item">02 datasets</a>
</li>
                                    
<li>
    <a href="../../../01_database/03_dbNSFP/" class="dropdown-item">03 dbNSFP</a>
</li>
                                    
<li>
    <a href="../../../01_database/04_dbSNP/" class="dropdown-item">04 dbSNP</a>
</li>
                                    
<li>
    <a href="../../../01_database/05_MalaCards/" class="dropdown-item">05 MalaCards</a>
</li>
                                    
<li>
    <a href="../../../01_database/06_revel/" class="dropdown-item">06 revel</a>
</li>
                                    
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">Pic</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../../01_database/pic/readme/" class="dropdown-item">Readme</a>
</li>
    </ul>
  </li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">02 course <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../02_course/01_%E7%94%9F%E4%BF%A1%E8%B5%84%E6%BA%90%E8%AF%BE/" class="dropdown-item">01 生信资源课</a>
</li>
                                    
<li>
    <a href="../../../02_course/02_%E6%B5%8B%E5%BA%8F%E5%8E%9F%E7%90%86/" class="dropdown-item">02 测序原理</a>
</li>
                                    
<li>
    <a href="../../../02_course/Agilent_SureSelect/" class="dropdown-item">Agilent SureSelect</a>
</li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">04 linuxshell <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../04_linuxshell/0/" class="dropdown-item">0</a>
</li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">07 microbe <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../07_microbe/00_%E7%BB%BC%E8%BF%B0/" class="dropdown-item">00 综述</a>
</li>
                                    
<li>
    <a href="../../../07_microbe/01_%E6%8B%BC%E6%8E%A5%E7%BB%84%E8%A3%85/" class="dropdown-item">IDBA-UD：用于深度非常不均匀的单细胞和宏基因组测序数据的组装软件</a>
</li>
                                    
<li>
    <a href="../../../07_microbe/02_%E5%AE%8F%E7%97%85%E6%AF%92%E7%BB%84/" class="dropdown-item">02 宏病毒组</a>
</li>
                                    
<li>
    <a href="../../../07_microbe/03_nCov19%E8%B5%84%E6%BA%90/" class="dropdown-item">03 nCov19资源</a>
</li>
                                </ul>
                            </li>
                            <li class="dropdown active">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">08 DNA <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../002_BAF/" class="dropdown-item">002 BAF</a>
</li>
                                    
<li>
    <a href="../../003_VEP/" class="dropdown-item">003 VEP</a>
</li>
                                    
<li>
    <a href="../../003_annovar/" class="dropdown-item">003 annovar</a>
</li>
                                    
<li>
    <a href="../../004_benchmark-datasets/" class="dropdown-item">数据集</a>
</li>
                                    
<li>
    <a href="../../006_sam_bam_samtools/" class="dropdown-item">006 sam bam samtools</a>
</li>
                                    
<li>
    <a href="../../01_HGVS/" class="dropdown-item">01 HGVS</a>
</li>
                                    
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">001 NIPT</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../001_NIPT/00/" class="dropdown-item">00</a>
</li>
    </ul>
  </li>
                                    
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">004 TCGA</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../004_TCGA/00/" class="dropdown-item">00</a>
</li>
            
<li>
    <a href="../../004_TCGA/01_%E7%BB%93%E7%9B%B4%E8%82%A0%E7%99%8C/" class="dropdown-item">01 结直肠癌</a>
</li>
    </ul>
  </li>
                                    
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">05 somatic</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/001_Mutect2/" class="dropdown-item">001 Mutect2</a>
</li>
            
<li>
    <a href="../../05_somatic/007_Tumor-Heterogeneity-Analysis/" class="dropdown-item">007 Tumor Heterogeneity Analysis</a>
</li>
            
<li>
    <a href="../../05_somatic/008_cfdna/" class="dropdown-item">008 cfdna</a>
</li>
            
<li>
    <a href="../../05_somatic/008_cfdna_paper/" class="dropdown-item">https://github.com/dkleftogi/cfDNA_AnalysisPipeline</a>
</li>
            
<li>
    <a href="../../05_somatic/009_MSI/" class="dropdown-item">009 MSI</a>
</li>
            
<li>
    <a href="../../05_somatic/009_MSI_paper/" class="dropdown-item">009 MSI paper</a>
</li>
            
<li>
    <a href="../../05_somatic/009_%E4%B8%B2%E8%81%94%E9%87%8D%E5%A4%8D%E5%BA%8F%E5%88%97/" class="dropdown-item">009 串联重复序列</a>
</li>
            
<li>
    <a href="../../05_somatic/010_drivergene/" class="dropdown-item">010 drivergene</a>
</li>
            
<li>
    <a href="../../05_somatic/014_MAF_maftools/" class="dropdown-item">014 MAF maftools</a>
</li>
            
<li>
    <a href="../../05_somatic/014_TMB/" class="dropdown-item">014 TMB</a>
</li>
            
<li>
    <a href="../../05_somatic/015_STAR/" class="dropdown-item">015 STAR</a>
</li>
            
<li>
    <a href="../../05_somatic/016.drivegene/" class="dropdown-item">https://software.broadinstitute.org/cancer/cga/mutsig_faq</a>
</li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">000 癌</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/000_%E7%99%8C/000_/" class="dropdown-item">000 </a>
</li>
            
<li>
    <a href="../../05_somatic/000_%E7%99%8C/001_%E6%98%93%E6%84%9F%E5%9F%BA%E5%9B%A0/" class="dropdown-item">001 易感基因</a>
</li>
    </ul>
  </li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">001 SNVINDDEL</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/001_SNVINDDEL/001_Notes_on_Mutect2/" class="dropdown-item">001 Notes on Mutect2</a>
</li>
            
<li>
    <a href="../../05_somatic/001_SNVINDDEL/001_paper/" class="dropdown-item">001 paper</a>
</li>
            
<li>
    <a href="../../05_somatic/001_SNVINDDEL/001_somatic_slides/" class="dropdown-item">MuTect</a>
</li>
            
<li>
    <a href="../../05_somatic/001_SNVINDDEL/002_MuTect/" class="dropdown-item">002 MuTect</a>
</li>
            
<li>
    <a href="../../05_somatic/001_SNVINDDEL/004_Pisces/" class="dropdown-item">004 Pisces</a>
</li>
            
<li>
    <a href="../../05_somatic/001_SNVINDDEL/004_VarDict/" class="dropdown-item">004 VarDict</a>
</li>
            
<li>
    <a href="../../05_somatic/001_SNVINDDEL/007_panel-of-normals/" class="dropdown-item">007 panel of normals</a>
</li>
    </ul>
  </li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">002 SV</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/002_SV/00_paper/" class="dropdown-item">00 paper</a>
</li>
            
<li>
    <a href="../../05_somatic/002_SV/00_%E7%BB%BC%E8%BF%B0/" class="dropdown-item">00 综述</a>
</li>
            
<li>
    <a href="../../05_somatic/002_SV/01_manta/" class="dropdown-item">01 manta</a>
</li>
            
<li>
    <a href="../../05_somatic/002_SV/02_AnnotSV/" class="dropdown-item">02 AnnotSV</a>
</li>
            
<li>
    <a href="../../05_somatic/002_SV/03_lumpy_svtyper/" class="dropdown-item">03 lumpy svtyper</a>
</li>
            
<li>
    <a href="../../05_somatic/002_SV/04_delly/" class="dropdown-item">04 delly</a>
</li>
    </ul>
  </li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">003 CNV</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/003_CNV/001_GATKCNV/" class="dropdown-item">001 GATKCNV</a>
</li>
            
<li>
    <a href="../../05_somatic/003_CNV/005_CNV_CNVkit/" class="dropdown-item">005 CNV CNVkit</a>
</li>
            
<li>
    <a href="../../05_somatic/003_CNV/005_CNV_DeCoN/" class="dropdown-item">005 CNV DeCoN</a>
</li>
            
<li>
    <a href="../../05_somatic/003_CNV/005_CNV_gistic/" class="dropdown-item">005 CNV gistic</a>
</li>
            
<li>
    <a href="../../05_somatic/003_CNV/005_CNV_paper/" class="dropdown-item">A Comprehensive Workflow for Read Depth-Based Identification of Copy-Number Variation from Whole-Genome Sequence Data</a>
</li>
            
<li>
    <a href="../../05_somatic/003_CNV/005_CNV_purecn/" class="dropdown-item">005 CNV purecn</a>
</li>
            
<li>
    <a href="../../05_somatic/003_CNV/00_cnv/" class="dropdown-item">历史背景</a>
</li>
            
<li>
    <a href="../../05_somatic/003_CNV/01_facets/" class="dropdown-item">01 facets</a>
</li>
            
<li>
    <a href="../../05_somatic/003_CNV/02_%20Canvas/" class="dropdown-item">02  Canvas</a>
</li>
            
<li>
    <a href="../../05_somatic/003_CNV/07_Canvas/" class="dropdown-item">07 Canvas</a>
</li>
            
<li>
    <a href="../../05_somatic/003_CNV/08_PennCNV/" class="dropdown-item">08 PennCNV</a>
</li>
    </ul>
  </li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">005 MSI</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/005_MSI/01_/" class="dropdown-item">01 </a>
</li>
    </ul>
  </li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">008 cfdna</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/008_cfdna/00/" class="dropdown-item">00</a>
</li>
    </ul>
  </li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">009 CellularityAndPloidy</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/009_CellularityAndPloidy/00_%E4%BB%8B%E7%BB%8D/" class="dropdown-item">00 介绍</a>
</li>
            
<li>
    <a href="../../05_somatic/009_CellularityAndPloidy/01_sequenza/" class="dropdown-item">介绍</a>
</li>
            
<li>
    <a href="../../05_somatic/009_CellularityAndPloidy/02_ABSOLUTE/" class="dropdown-item">02 ABSOLUTE</a>
</li>
    </ul>
  </li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">010 HLA</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/010_HLA/001_polysolver/" class="dropdown-item">001 polysolver</a>
</li>
            
<li>
    <a href="../../05_somatic/010_HLA/003_HLAreporter/" class="dropdown-item">003 HLAreporter</a>
</li>
            
<li>
    <a href="../../05_somatic/010_HLA/00_%E7%AE%80%E4%BB%8B/" class="dropdown-item">00 简介</a>
</li>
            
<li>
    <a href="../../05_somatic/010_HLA/010_HLA/" class="dropdown-item">010 HLA</a>
</li>
            
<li>
    <a href="../../05_somatic/010_HLA/010_seq2HLA/" class="dropdown-item">010 seq2HLA</a>
</li>
    </ul>
  </li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">011 NIPT</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/011_NIPT/00/" class="dropdown-item">00</a>
</li>
    </ul>
  </li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">012 mutation sig</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/012_mutation_sig/00_%E7%AE%80%E4%BB%8B/" class="dropdown-item">00 简介</a>
</li>
            
<li>
    <a href="../../05_somatic/012_mutation_sig/01_MutationalPattern/" class="dropdown-item">01 MutationalPattern</a>
</li>
            
<li>
    <a href="../../05_somatic/012_mutation_sig/02_SigProfilerMatrixGenerator/" class="dropdown-item">02 SigProfilerMatrixGenerator</a>
</li>
            
<li>
    <a href="../../05_somatic/012_mutation_sig/03_superFreq/" class="dropdown-item">03 superFreq</a>
</li>
            
<li>
    <a href="../../05_somatic/012_mutation_sig/04_deconstructSigs/" class="dropdown-item">04 deconstructSigs</a>
</li>
    </ul>
  </li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">013 mutation spectrum</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/013_mutation_spectrum/00/" class="dropdown-item">00</a>
</li>
    </ul>
  </li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">014 fusion</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/014_fusion/000_%E7%BB%BC%E8%BF%B0/" class="dropdown-item">000 综述</a>
</li>
            
<li>
    <a href="../../05_somatic/014_fusion/014_STARfusion/" class="dropdown-item">014 STARfusion</a>
</li>
            
<li>
    <a href="../../05_somatic/014_fusion/014_gFusion/" class="dropdown-item">014 gFusion</a>
</li>
    </ul>
  </li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">016 克隆进化</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../05_somatic/016_%E5%85%8B%E9%9A%86%E8%BF%9B%E5%8C%96/00/" class="dropdown-item">00</a>
</li>
    </ul>
  </li>
    </ul>
  </li>
                                    
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">06 germline</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="./" class="dropdown-item active">VCF</a>
</li>
            
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">015</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../015/1/" class="dropdown-item">1</a>
</li>
    </ul>
  </li>
    </ul>
  </li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">09 scrna <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../09_scrna/00/" class="dropdown-item">00</a>
</li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">10 NGS <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../10_NGS/000_linux/" class="dropdown-item">000 linux</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/000_linuxdemo/" class="dropdown-item">000 linuxdemo</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/001_SRA/" class="dropdown-item">001 SRA</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/002_%E5%BC%95%E7%89%A9%E8%AE%BE%E8%AE%A1/" class="dropdown-item">002 引物设计</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/003_shell/" class="dropdown-item">003 shell</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/004_genome/" class="dropdown-item">004 genome</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/005_docker/" class="dropdown-item">005 docker</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/005_docker%E8%BF%9B%E9%98%B6/" class="dropdown-item">005 docker进阶</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/006_git/" class="dropdown-item">006 git</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/007_ack/" class="dropdown-item">007 ack</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/008_tmux/" class="dropdown-item">008 tmux</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/009_NCBI/" class="dropdown-item">api</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/010_Allele%20Balance/" class="dropdown-item">010 Allele Balance</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/011_ORF/" class="dropdown-item">011 ORF</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/012_%E4%B8%8B%E8%BD%BD/" class="dropdown-item">012 下载</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/013_bed/" class="dropdown-item">013 bed</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/014_blat/" class="dropdown-item">014 blat</a>
</li>
                                    
<li>
    <a href="../../../10_NGS/03/" class="dropdown-item">03</a>
</li>
                                    
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">01 QC</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../../10_NGS/01_QC/01_cutadapt/" class="dropdown-item">01 cutadapt</a>
</li>
            
<li>
    <a href="../../../10_NGS/01_QC/05_fastp/" class="dropdown-item">05 fastp</a>
</li>
    </ul>
  </li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">11 ML <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../11_ML/001_/" class="dropdown-item">001 </a>
</li>
                                    
<li>
    <a href="../../../11_ML/%E8%B4%9D%E5%8F%B6%E6%96%AF/" class="dropdown-item">贝叶斯</a>
</li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">12 RNA <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../12_RNA/00/" class="dropdown-item">00</a>
</li>
                                    
  <li class="dropdown-submenu">
    <a href="#" class="dropdown-item">miRNA</a>
    <ul class="dropdown-menu">
            
<li>
    <a href="../../../12_RNA/miRNA/01/" class="dropdown-item">01</a>
</li>
    </ul>
  </li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">13 R PERL wdl <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../13_R_PERL_wdl/00.PERL/" class="dropdown-item">00.PERL</a>
</li>
                                    
<li>
    <a href="../../../13_R_PERL_wdl/00.R/" class="dropdown-item">00.R</a>
</li>
                                    
<li>
    <a href="../../../13_R_PERL_wdl/00.WDL/" class="dropdown-item">00.WDL</a>
</li>
                                    
<li>
    <a href="../../../13_R_PERL_wdl/001_%E8%B4%9D%E5%8F%B6%E6%96%AF/" class="dropdown-item">001 贝叶斯</a>
</li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">14 immune <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../14_immune/00/" class="dropdown-item">00</a>
</li>
                                    
<li>
    <a href="../../../14_immune/01_%20pVACtools/" class="dropdown-item">01  pVACtools</a>
</li>
                                    
<li>
    <a href="../../../14_immune/01_IMGT/" class="dropdown-item">01 IMGT</a>
</li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">15 python <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../15_python/000_course/" class="dropdown-item">000 course</a>
</li>
                                    
<li>
    <a href="../../../15_python/001_pandas/" class="dropdown-item">001 pandas</a>
</li>
                                    
<li>
    <a href="../../../15_python/002_xls/" class="dropdown-item">002 xls</a>
</li>
                                    
<li>
    <a href="../../../15_python/003_conda/" class="dropdown-item">003 conda</a>
</li>
                                    
<li>
    <a href="../../../15_python/004_%E6%80%A7%E8%83%BD%E5%88%86%E6%9E%90%E5%99%A8/" class="dropdown-item">004 性能分析器</a>
</li>
                                    
<li>
    <a href="../../../15_python/005_request/" class="dropdown-item">005 request</a>
</li>
                                    
<li>
    <a href="../../../15_python/005_%E7%88%AC%E8%99%AB/" class="dropdown-item">Robots协议</a>
</li>
                                    
<li>
    <a href="../../../15_python/007_parallel/" class="dropdown-item">007 parallel</a>
</li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">16 TGS <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../16_TGS/001/" class="dropdown-item">Nanopore</a>
</li>
                                    
<li>
    <a href="../../../16_TGS/002_minimap2/" class="dropdown-item">002 minimap2</a>
</li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">Slides <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li>
    <a href="../../../slides/20191121/" class="dropdown-item">20191121</a>
</li>
                                    
<li>
    <a href="../../../slides/20191122/" class="dropdown-item">20191122</a>
</li>
                                    
<li>
    <a href="../../../slides/20191128/" class="dropdown-item">Blood-based tumor mutational burden as a predictor of clinical benefit in non-small-cell lung cancer patients treated with atezolizumab</a>
</li>
                                    
<li>
    <a href="../../../slides/20200130/" class="dropdown-item">Enhancing sensitivity and controlling false discovery rate in somatic indel discovery</a>
</li>
                                    
<li>
    <a href="../../../slides/cfnda-protocl/" class="dropdown-item">Cfnda protocl</a>
</li>
                                    
<li>
    <a href="../../../slides/cosmic/" class="dropdown-item">Cosmic</a>
</li>
                                    
<li>
    <a href="../../../slides/marp/" class="dropdown-item">Marp</a>
</li>
                                    
<li>
    <a href="../../../slides/umi/" class="dropdown-item">Umi</a>
</li>
                                    
<li>
    <a href="../../../slides/%E7%97%85%E6%AF%92%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/" class="dropdown-item">病毒宏基因组</a>
</li>
                                </ul>
                            </li>
                        </ul>

                    <ul class="nav navbar-nav ml-auto">
                        <li class="nav-item">
                            <a href="#" class="nav-link" data-toggle="modal" data-target="#mkdocs_search_modal">
                                <i class="fa fa-search"></i> Search
                            </a>
                        </li>
                            <li class="nav-item">
                                <a rel="prev" href="../../05_somatic/016_%E5%85%8B%E9%9A%86%E8%BF%9B%E5%8C%96/00/" class="nav-link">
                                    <i class="fa fa-arrow-left"></i> Previous
                                </a>
                            </li>
                            <li class="nav-item">
                                <a rel="next" href="../015/1/" class="nav-link">
                                    Next <i class="fa fa-arrow-right"></i>
                                </a>
                            </li>
                    </ul>
                </div>
            </div>
        </div>

        <div class="container">
            <div class="row">
                    <div class="col-md-3"><div class="navbar-light navbar-expand-md bs-sidebar hidden-print affix" role="complementary">
    <div class="navbar-header">
        <button type="button" class="navbar-toggler collapsed" data-toggle="collapse" data-target="#toc-collapse" title="Table of Contents">
            <span class="fa fa-angle-down"></span>
        </button>
    </div>

    
    <div id="toc-collapse" class="navbar-collapse collapse card bg-secondary">
        <ul class="nav flex-column">
            
            <li class="nav-item" data-level="1"><a href="#vcf" class="nav-link">VCF</a>
              <ul class="nav flex-column">
            <li class="nav-item" data-level="2"><a href="#local-realignment" class="nav-link">Local Realignment</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#local-reaglinment" class="nav-link">Local reaglinment</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#duplicate-markingremoval" class="nav-link">Duplicate marking/removal</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#duplicate-reads-fragments" class="nav-link">Duplicate reads (fragments)</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#base-quality-score-recalibration" class="nav-link">Base quality score recalibration</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#bqsr" class="nav-link">BQSR</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#variant-calling-genotyping" class="nav-link">Variant calling &amp; genotyping</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#variant-filtration-soft" class="nav-link">Variant filtration (soft)</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#annotation" class="nav-link">Annotation</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#genotyping" class="nav-link">Genotyping:</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#haplotypes-and-computational-phasing" class="nav-link">Haplotypes and Computational Phasing</a>
              <ul class="nav flex-column">
              </ul>
            </li>
              </ul>
            </li>
        </ul>
    </div>
</div></div>
                    <div class="col-md-9" role="main">

<h1 id="vcf">VCF</h1>
<pre><code>ID: some ID for the variant, if known (e.g., dbSNP)
REF, ALT: reference and alternative alleles (on forward strand of reference)
QUAL = -10*log(1-p), where p is the probability of variant being present given the read data
FILTER: whether the variant failed a filter (filters defined by the user or program processing the file)
GT (genotype):
  0/0 reference homozygote
  0/1 reference-alternative heterozygote
  1/1 alternative homozygote
  0/2, 1/2, 2/2, etc. - possible if more than one alternative allele present
  ./. missing data
AD: allele depths
DP: total depth (may be different from sum of AD depths, a the latter include only reads significantly supporting alleles)
PL: genotype likelihoods (phred-scaled), normalized to the best genotype, e.g., PL(0/1) = -10*log[ Prob(data|0/1) 
/ Prob(data|best_genotype) ]
GQ: genotype quality – this is just PL of the second-best genotype
</code></pre>

<p>QUAL：基于Phred格式的表示ALT的质量，也可以理解为可靠性；可以理解为所call出来的变异位点的质量值。Q=-10lgP，Q表示质量值；P表示这个位点发生错误的概率。因此，如果想把错误率从控制在90%以上，P的阈值就是1/10，那lg（1/10）=-1，Q=（-10）*（-1）=10。同理，当Q=20时，错误率就控制在了0.01。这个参数其实和mpileup的-Q是重复的。</p>
<p>MQ不同于MAPQ，是RMS Mapping Quality，公式定义如下：q指的是比对到这个参考基因组碱基上的比对质量，即MAPQ。参考之前说的，MAPQ设置为20，假设每个碱基的MAPQ都是20，则MQ为20。</p>
<p>$$RMS=\sqrt{\frac{1}{N}\sum_{n}^{i=1}q_{i}^{2}}$$</p>
<p>大多数的reads的MAPQ都在60。</p>
<h2 id="local-realignment">Local Realignment</h2>
<p>Why?
+ Each read is independently aligned to the reference genome to cut down
on the computational cost.
+ But this means that many reads spanning indels are likely to be
misaligned.
+ We see this as regions containing indels as well as clusters of
mismatching bases.</p>
<p>How does it work?
+ Step 1: Identify regions likely in need of realignment.
  + 1 – At least one read contains an indel.
  + 2 – A cluster of mismatching bases exists.
  + 3 – An already known indel segregates at the site.
+ Step 2: Perform a multiple sequence realignment in this region, such that the number of mismatching bases is minimized across all reads.</p>
<h2 id="local-reaglinment">Local reaglinment</h2>
<p><img alt="xx" src="https://github.com/xiucz/pics/blob/master/20191027.jpg?raw=true" /></p>
<p>http://www.cbs.dtu.dk/courses/27626/Lectures/post_alignment_variantcalling_SR.pdf</p>
<h3 id="ambiguity-of-alignment-at-indel-sites">Ambiguity of alignment at indel sites</h3>
<p><img alt="xx" src="https://github.com/xiucz/pics/blob/master/201910272.jpg?raw=true" /></p>
<p>Re-alignment no longer recommended if the genotyping method used downstream involves local haplotype 
assembly
FreeBayes，re-alignment implicit in the assembly algorithm；
HaplotypeCaller</p>
<p>Still needed if genotypes called from allelic depths at individual sites:
UnifiedGenotyper (GATK)
samtools</p>
<hr />
<h2 id="duplicate-markingremoval">Duplicate marking/removal</h2>
<p>Why do we have duplicates in our data?
• The PCR amplification step included in the majority of NGS library construction techniques can introduce duplicates in the data.</p>
<p>Why do we need to remove these?
+ It will bias our variant calls.
+ PCR errors are propagated and sampled many times giving rise to FP variant calls.</p>
<p>Basic concepts of duplicate marking
http://picard.sourceforge.net/</p>
<p>algorithm:
1. Identify genomic position and strand for 5’-most bases.
2. Mark reads that are duplicatesof each other.
3. Within a group of duplicate reads the read with the highest sum of base quality scores is retained.</p>
<p>But it is not perfect…
+ Does not account for sequencing errors.
+ Does not account for natural duplicates.
+ Does not account for duplicate reads with different mapping locations</p>
<h2 id="duplicate-reads-fragments">Duplicate reads (fragments)</h2>
<p>Optical duplicates: (Illumina) generated when a single cluster of reads is part of two adjacent tiles' on the same 
slide and used to compute two read calls separately
• Very similar in sequence (except sequencing errors).
• Identified where the first, say, 50 bases are identical between two reads and the read’s coordinates are close</p>
<p>Library duplicates (aka PCR duplicates): generated when the original sample is preamplified to such extent 
that initial unique targets are PCR replicated prior to library preparation and will lead to several independent 
spots on the Illumina slide.
• do not have to be adjacent on the slide
• share a very high level of sequence identity
• align to the same place on reference
• identified from alignment to reference </p>
<p><img alt="xx" src="https://github.com/xiucz/pics/blob/master/201910271.jpg?raw=true" /></p>
<h3 id="removing-marking-duplicates-with-picard">Removing (marking) duplicates with PICARD</h3>
<pre><code>java -jar $PICARDDIR/picard.jar \
MarkDuplicates \
INPUT=sample1.sorted.bam \
OUTPUT=sample1.sorted.dedup.bam \
METRICS_FILE=sample1.sorted.dedup.metrics.txt
</code></pre>

<p>The metrics file will contain some stats about the de-duping
In the resulting BAM file, only one fragment from each duplicate group survives unchanged, other duplicate 
      fragments are given a flag 0x400 and will not be used downstream. 
Optimally, detection and marking of duplicate fragments should be done per library, i.e., over all read groups 
corresponding to a given library. • In practice, often sufficient to do it per lane (read group).</p>
<hr />
<h2 id="base-quality-score-recalibration">Base quality score recalibration</h2>
<p>How it works!
1. Collect information regarding the following features of the bases:
  + Reported quality score
  + Position within the read (machine cycle)
  + Dinucleotide context (current base plus previous base)
2. Count the number of times a site was a mismatch to the reference (excluding known polymorphic sites).
3. Estimate new quality score as:</p>
<pre><code>$$Phred-scaled quality score =\frac{# of reference mismatches +1}{# of observed bases +2}$$
</code></pre>

<p>Example:
We observed A [AA, pos. 35, Q20] a 1,000,000 times.A in this context mismatches the reference a 1,000 times. This gives us: </p>
<pre><code>Q value = -10log10((1,000+1)/(1,000,000+2)) ~ Q30
</code></pre>

<h2 id="bqsr">BQSR</h2>
<p>Caveats: 
Local indel realignment should be done before recalibration</p>
<p>Known variation (SNPs and indels) have to be excluded (not a source of errors)</p>
<p>Variants found from read pileup; PL from allelic read depths and base qualities</p>
<p>Variants found from locally assembled haplotypes; PL from read alignment to haplotypes</p>
<h2 id="variant-calling-genotyping">Variant calling &amp; genotyping</h2>
<p>The posterior probability of each genotype given the pileup of sequence reads is given by Bayes theorem:</p>
<p><img alt="xx" src="https://github.com/xiucz/pics/blob/master/201910276.jpg?raw=true" /></p>
<ul>
<li>p(G|D): Prob. of genotype given data (posterior probability)</li>
<li>p(G): Prior prob. based on allele freqs. from HW</li>
<li>p(D|G): Prob. of data given genotype (genotype likelihood)</li>
<li><strong>p(D):</strong> Prob. of data</li>
<li>p(b|G): Prob. of genotype given observed base at a site</li>
<li>e is the Phred scaled quality score of the base</li>
</ul>
<p><img alt="xx" src="https://github.com/xiucz/pics/blob/master/201910277.jpg?raw=true" />
<img alt="xx" src="https://github.com/xiucz/pics/blob/master/201910278.jpg?raw=true" />
<img alt="xx" src="https://github.com/xiucz/pics/blob/master/201910279.jpg?raw=true" /></p>
<h2 id="variant-filtration-soft">Variant filtration (soft)</h2>
<p><img alt="xx" src="https://github.com/xiucz/pics/blob/master/2019102710.jpg?raw=true" /></p>
<h3 id="variant-quality-score-recalibration">Variant quality score recalibration</h3>
<ul>
<li>Variants as points in a cloud can be modeled using a Gaussian mixture model.</li>
<li>This model is used to estimate the probability that a variant is a true variant rather than a sequencer error or a data processing artifact.</li>
<li>The model is trained on known ‘true sites’.</li>
<li>Model parameters are applied to each variant in the input vcf-file and each variant is flagged by a VQSLOD score, which is the log odds ratio of the variant being a true variant versus being a false variant under the trained Gaussian mixture model. </li>
</ul>
<h2 id="annotation">Annotation</h2>
<p><img alt="xx" src="https://github.com/xiucz/pics/blob/master/2019102711.jpg?raw=true" /></p>
<h2 id="genotyping">Genotyping:</h2>
<h3 id="maximum-likelihood-estimator">Maximum likelihood estimator</h3>
<p>假设在某个总深度为90条reads的SNP位点观察到25条突变reads，该怎么估计该位点的基因型呢？</p>
<p>三种可能的基因型：
+ RR  <br />
+ RA  <br />
+ AA    </p>
<p>计算似然值，即某个指定基因型观察值的可能性
+ P(25AltReads,65RefReads|RR) <br />
+ P(25AltReads,65RefReads|RA) <br />
+ P(25AltReads,65RefReads|AA)</p>
<p>通常，我们观察到突变reads数目服从二项分布：</p>
<p>$$L(x|G) =\binom{90}{25}p^{25}(1-p)^{^{25}}$$</p>
<p>p的函数，假设该位点是二倍体，则p有3种可能的值：
+ RR    -&gt;  p=0 
+ RA    -&gt;  p=0.5 <br />
+ AA    -&gt;  p=1 </p>
<p>由于要考虑到比对中的技术错误：
+ RR    -&gt;  p=0.05<br />
+ RA    -&gt;  p=0.5 <br />
+ AA    -&gt;  p=0.95</p>
<h3 id="incorporating-additional-information-in-our-model">Incorporating additional information in our model</h3>
<p><strong>Bayesian  Inference</strong></p>
<h3 id="bayesian-statistics-vs-frequentist-statistics">Bayesian Statistics vs. Frequentist Statistics</h3>
<ul>
<li>The   main problem of Statistcal Inference is the estimation of a specific feature (parameter) of a population    (for example, the genotype of a given individual on a given locus)  </li>
<li>Classic (Frequentist) Statistics assume that the feature is deterministic(fixed at the time of our observation)   </li>
<li>Bayesian  Statistics  assume  that    the feature is  random      </li>
<li>This  (apparently)    small   technical   difference  creates a   huge    philosophical   dis;nc;on <br />
(and    a   lot of  technicali;es   too)    </li>
<li>Sta;s;cians   tend    to  be  very    passionate  around  each    approach    (although   some    are 
pragma;c)</li>
</ul>
<h3 id="bayesian-statistics">Bayesian Statistics</h3>
<p>• If    the feature of  the popula;on   that    we  want    to  es;mate is  a   random  variable,   it<br />
must    have    a   probability distribu;on.    We  call    this    the prior   distribu9on:    P(G)      <br />
• We    take    a   sample  from    the popula;on,  measure our observa;ons,    and compute 
the likelihood  or  our data:   P(R|G)<br />
• We    can then    update  our knowledge   of  the distribu;on of  the parameter   and obtain<br />
its posterior   distribu9on using   the Bayes   Theorem:  <br />
P(G | R) =
P(G)P(R |G)
P(R)
• P(R)  is  difficult   to  compute and it  is  the same    for all possible    genotypes,  so  in  most  <br />
applica;ons we  just    need    to  compute the numerator.<br />
• We    obtain  a   probability distribu;on!    Not a   single  value!  But we  can obtain  a   range <br />
of  values  with    high    probability (credible   interval),  or  summarise   the distribu;on 
with    the most    likely  value   (maximum    a   posteriori) or  the mean    (posterior  mean)   </p>
<h3 id="naive-bayesian-genotyping-posterior-probability">Naïve Bayesian genotyping: Posterior probability</h3>
<p>$$L(x|G) =\binom{90}{25}p^{25}(1-p)^{^{25}}$$</p>
<p>We  need    to  consider    Prior   probabili;es    for each    genotype: <br />
+ $P(RR)=0.75$
+ $P(RA)=0.15$
+ $P(AA)=0.10$</p>
<p>We  can compute posterior   probabili;es: <br />
+ P(RR|x)=P(x|RR)P(RR)<br />
+ P(RA|x)=P(x|RA)P(RA)<br />
+ P(AA|x)=P(x|AA)P(AA)  </p>
<p>In situations where we  don’t have much data (lower depth), or there is more noise, the prior can have more impact in the posterior. Let’s assume now 10 variant reads out of 50:</p>
<p><img alt="xx" src="https://github.com/xiucz/pics/blob/master/201910273.jpg?raw=true" />
<img alt="xx" src="https://github.com/xiucz/pics/blob/master/201910274.jpg?raw=true" />
<img alt="xx" src="https://github.com/xiucz/pics/blob/master/201910274.jpg?raw=true" /></p>
<h3 id="haplotype-likelihood-function-in-haplotypecaller">Haplotype likelihood function in HaplotypeCaller</h3>
<ol>
<li>Define active regions
The program determines which regions of the genome it needs to operate on, based on the presence of significant
evidence for variation.</li>
<li>Determine haplotypes by re-assembly of the active region
For each ActiveRegion, the program builds a De Bruijn-like graph to reassemble the ActiveRegion, and identifies what
are the possible haplotypes present in the data. The program then realigns each haplotype against the reference
haplotype using the Smith-Waterman algorithm in order to identify potentially variant sites.</li>
<li>Determine likelihoods of the haplotypes given the read data P{Dj|H}
For each ActiveRegion, the program performs a pairwise alignment of each read against each haplotype using the
PairHMM algorithm. This produces a matrix of likelihoods of haplotypes given the read data. These likelihoods are
then marginalized to obtain the likelihoods of alleles for each potentially variant site given the read data.</li>
<li>Assign sample genotypes
For each potentially variant site, the program applies Bayes’ rule, using the likelihoods of alleles given the read data
to calculate the likelihoods of each genotype per sample given the read data observed for that sample. The most
likely genotype is then assigned to the sample.</li>
</ol>
<p>Haplotype assembly uses reads from all samples (rather than one at a time), and so…resulting VCF file will not
 be exactly equivalent to that obtained from gVCF mode runs followed by GenotypeGVCF</p>
<p>Broad recommends running HaplotypeCaller-based pipeline instead of UnifiedGenotyper
However, still recommended for
Pooled sample cases
High ploidy cases</p>
<p>-stand_call_conf [number]
Variants with quality score (QUAL) less than [number] will not be written to VCF file. Good to set this
low – better have too many raw variants than too few. Can always filter VCF file later. Default: 30.</p>
<p>-dcov [int]
Read depth at any locus will be capped at [number]; the goal is to provide even distribution of read
start positions while removing excess coverage. For truly unbiased down-sampling, use -dfrac
Defaults are usually high (250) – can be reduced to speed things up</p>
<h2 id="haplotypes-and-computational-phasing">Haplotypes and Computational Phasing</h2>
<p>A   haplotype   is  a   set of  gene;c  features    that    tend    to  be  inherited   together  <br />
In  our context,    we  are interested  in  a   set of  SNPs    that    tend    to  be  inherited   together.     <br />
Why?    Because it  will    make    variant calling more    robust  </p></div>
            </div>
        </div>

        <footer class="col-md-12">
            <hr>
            <p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>
        <script>
            var base_url = "../../..",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
        <script src="../../../js/base.js" defer></script>
        <script src="../../../search/main.js" defer></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="searchModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-lg">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="searchModalLabel">Search</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
                <p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
                <form>
                    <div class="form-group">
                        <input type="search" class="form-control" placeholder="Search..." id="mkdocs-search-query" title="Type search term here">
                    </div>
                </form>
                <div id="mkdocs-search-results"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div><div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>

    </body>
</html>
